notify/templates.go

```diff
@@ -0,0 +1,240 @@
+package notify
+
+import (
+	"context"
+	"fmt"
+	tmplhtml "html/template"
+	"net/url"
+	"path/filepath"
+	tmpltext "text/template"
+	"text/template/parse"
+
+	"github.com/grafana/alerting/templates"
+	v2 "github.com/prometheus/alertmanager/api/v2"
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/template"
+	"github.com/prometheus/common/model"
+)
+
+type TestTemplatesConfigBodyParams struct {
+	// Alerts to use as data when testing the template.
+	Alerts []*PostableAlert `json:"alerts"`
+
+	// Template string to test.
+	Template string `json:"template"`
+
+	// Name of the template file.
+	Name string `json:"name"`
+}
+
+type TestTemplatesResults struct {
+	Results []TestTemplatesResult      `json:"results"`
+	Errors  []TestTemplatesErrorResult `json:"errors,omitempty"`
+}
+
+type TestTemplatesResult struct {
+	// Name of the associated template definition for this result.
+	Name string `json:"name"`
+
+	// Interpolated value of the template.
+	Text string `json:"text"`
+}
+
+type TestTemplatesErrorResult struct {
+	// Name of the associated template for this error. Will be empty if the Kind is "invalid_template".
+	Name string `json:"name,omitempty"`
+
+	// Kind of template error that occurred.
+	Kind TemplateErrorKind `json:"kind"`
+
+	// Error message.
+	Message string `json:"message"`
+}
+
+type TemplateErrorKind string
+
+const (
+	InvalidTemplate TemplateErrorKind = "invalid_template"
+	ExecutionError  TemplateErrorKind = "execution_error"
+)
+
+// TestTemplate tests the given template string against the given alerts. Existing templates are used to provide context for the test.
+// If an existing template of the same filename as the one being tested is found, it will not be used as context.
+func (am *GrafanaAlertmanager) TestTemplate(ctx context.Context, c TestTemplatesConfigBodyParams) (*TestTemplatesResults, error) {
+	definitions, err := parseTestTemplate(c.Name, c.Template)
+	if err != nil {
+		return &TestTemplatesResults{
+			Errors: []TestTemplatesErrorResult{{
+				Kind:    InvalidTemplate,
+				Message: err.Error(),
+			}},
+		}, nil
+	}
+
+	// Recreate the current template without the definition blocks that are being tested. This is so that any blocks that were removed don't get defined.
+	paths := make([]string, 0)
+	for _, name := range am.templates {
+		if name == c.Name {
```

Comment:
I don't think this works as expected? The following template called **other** doesn't return test2 or test3.

```
{{ define "test2" }}
test2
{{ end }}

{{ define "test3" }}
test3
{{ end }}
```

Reply:
Just tested and it works for me. It did make me notice an issue with detecting the wrapper template though, that extra space in the middle means the wrapper has a node with empty space in it. I should check `parse.IsEmptyTree(def.Root)` instead of `def.Root==nil`.

Reply:
I'll check again as I copied this comment from the original draft PR to avoid it getting lost 🙂

Reply:
Still a bug for me 🙁 

```go
paths := make([]string, 0)
for _, name := range am.templates {
	fmt.Println("existing tmpl is", name)
	if name == c.Name {
		// Skip the existing template of the same name as we're going to parse the one for testing instead.
		continue
	}
	paths = append(paths, filepath.Join(am.workingDirectory, name))
}
```

```
existing tmpl is test1
existing tmpl is others
existing tmpl is __default__.tmpl
```

It doesn't print `test2` or `test3`:

<img width="698" alt="Screenshot 2023-04-27 at 10 44 17" src="https://user-images.githubusercontent.com/85952834/234825151-0e41c343-e686-4e6e-9ad8-6573db994fd0.png">

Reply:
In the screenshot you're printing the filename of existing persisted templates, not the name of the various definitions. Also, it won't know about any of the templates currently being tested, those aren't persisted.